include/forcing/NetCDFPerFeatureDataProvider.hpp

```diff
@@ -0,0 +1,464 @@
+
+            // if time att is not encoded 
+            // TODO determine how this should be handled
+            std::string time_unit_str;
+
+            if ( time_unit_att.isNull() )
+            {
+                time_unit = TIME_SECONDS;
+            }
+            else
+            {  
+                time_unit_att.getValues(time_unit_str);
+            }
+
+            if ( time_unit_str == "h" || time_unit_str == "hours ")
+                time_unit = TIME_HOURS;
+            else if ( time_unit_str == "m" || time_unit_str == "minuets" )
+                    time_unit = TIME_MINUETS;
+            else if ( time_unit_str ==  "s" || time_unit_str == "seconds" )
+                    time_unit = TIME_SECONDS;
+            else if ( time_unit_str ==  "ms" || time_unit_str == "miliseconds" )
+                    time_unit = TIME_SECONDS;
+            else if ( time_unit_str ==  "us" || time_unit_str == "microseconds" )
+                    time_unit = TIME_SECONDS;
+            else if ( time_unit_str ==  "ns" || time_unit_str == "nanoseconds" )
+                    time_unit = TIME_SECONDS;
+            
+            double time_scale_factor;
+
+            switch (time_unit)
+            {
+                case TIME_HOURS:
+                    time_scale_factor = 3600;
+                break;
+
+                case TIME_MINUETS:
+                    time_scale_factor = 60;
+                break;
+
+                case TIME_SECONDS:
+                    time_scale_factor = 1;
+                break;
+
+                case TIME_MILLISECONDS:
+                    time_scale_factor = .001;
+                break;
+
+                case TIME_MICROSECONDS:
+                    time_scale_factor = .000001;
+                break;
+
+                case TIME_NANOSECONDS:
+                    time_scale_factor = .000000001;
+                break;
+            }
+
+            // read the meta data to get the epoc start
+            auto epoc_att = time_var.getAtt("epoc_start");
+            std::string epoc_start_str;
+
+            if ( epoc_att.isNull() )
+            {
+                epoc_start_str = "01/01/1970 00:00:00";
+            }
+            else
+            {  
+                epoc_att.getValues(epoc_start_str);
+            }
+
+            std::tm tm{};
+            std::stringstream s(epoc_start_str);
+            s >> std::get_time(&tm, "%F %R");
+            std::time_t epoc_start_time = mktime(&tm);
+
+            // scale the time to account for time units and epoc_start
+            // TODO make sure this happens with a FMA instruction
+            time_vals.resize(raw_time.size());
+            std::transform(raw_time.begin(), raw_time.end(), time_vals.begin(), 
+                [&](const auto& n){return n * time_scale_factor + epoc_start_time; });
+                
+
+            // determine the stride of the time array
+            time_stride = time_vals[1] - time_vals[0];
```

Comment:
This assumes the data has a known frequency.  This may not always be the case...

Reply:
We need that measurement currently, do you see other options?

Reply:
My point here is that this implementation assumes the data occurs at a frequency that is consistent with the first two entries.  If there are cases where that assumption is used in other parts of the code, then the data should be validated for the assumption before applying it.

Reply:
Validation might prove to very expensive if the time dimension is large enough. We certainly would not want to keep revalidating the same file.

Reply:
Some options here:

1.  Add frequency to meta data and use that <preferred>
2.  Do some random sampling to see if there is consistent time step
3. Assume the time step is not consistent and use a binary_search to find the correct index positions (O(lg n) instead of O(1) )

It is worth notting that **3** still assumes that the time values are in order and increasing

Reply:
My thought: what is assumed here is going to be the most common use case. We should require some other signal that this assumption should not be used if we need to do something more expensive (metadata in NetCDF file? subclass `NonLinearTimeNetCdfPerFeatureDataProvider`? Flag in forcing configuration in realization config?). We could also do a minimal test to see if the assumption is likely incorrect, that is, see if delta from 0 to 1 times the number of timesteps matches the end time--if not, something is wrong (would not catch all cases where the assumption was false, but likely 99% of them).